plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java

```diff
@@ -0,0 +1,251 @@
+ */
+package com.bolyuba.nexus.plugin.npm.hosted;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.zip.GZIPInputStream;
+
+import org.sonatype.nexus.proxy.item.StorageCollectionItem;
+import org.sonatype.nexus.proxy.item.StorageFileItem;
+import org.sonatype.nexus.proxy.item.StorageItem;
+import org.sonatype.nexus.proxy.walker.AbstractWalkerProcessor;
+import org.sonatype.nexus.proxy.walker.WalkerContext;
+import org.sonatype.nexus.util.io.StreamSupport;
+import org.sonatype.sisu.goodies.common.Iso8601Date;
+
+import com.bolyuba.nexus.plugin.npm.internal.Version;
+import com.bolyuba.nexus.plugin.npm.service.HostedMetadataService;
+import com.bolyuba.nexus.plugin.npm.service.PackageRoot;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.Maps;
+import com.google.common.primitives.Ints;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * Processor that rebuilds/updates npm metadata from data extracted from tarballs.
+ *
+ * @since 2.11
+ */
+public class RecreateMetadataWalkerProcessor
+    extends AbstractWalkerProcessor
+{
+  private static final String PACKAGE_JSON_PATH = "/package.json";
+
+  private static final int MAX_PACKAGE_JSON_FILE_SIZE = 1_000_000;
+
+  private final Logger log = LoggerFactory.getLogger(getClass());
+
+  private final ObjectMapper objectMapper;
+
+  private final HostedMetadataService hostedMetadataService;
+
+  private int packageRoots;
+
+  private int packageVersions;
+
+  public RecreateMetadataWalkerProcessor(final HostedMetadataService hostedMetadataService)
+  {
+    this.objectMapper = new ObjectMapper();
+    this.hostedMetadataService = hostedMetadataService;
+    this.packageRoots = 0;
+    this.packageVersions = 0;
+  }
+
+  public int getPackageRoots() {
+    return packageRoots;
+  }
+
+  public int getPackageVersions() {
+    return packageVersions;
+  }
+
+  @Override
+  public void processItem(final WalkerContext context, final StorageItem item)
+      throws Exception
+  {
+    // nop
+  }
+
+  public void onCollectionExit(final WalkerContext context, final StorageCollectionItem coll)
+      throws Exception
+  {
+    final TreeMap<Version, Map<String, Object>> versions = Maps.newTreeMap();
+
+    // collect all the possible tarballs
+    final Collection<StorageItem> items = coll.list();
+    for (StorageItem item : items) {
+      if (!(item instanceof StorageFileItem)) {
+        // not a file, just skip it
+        continue;
+      }
+      if (!item.getName().endsWith(".tgz")) {
```

Comment:
Is this consistent for all tarballs? Not overly familiar with npm, but ".tar.gz" is equivalent